### PR TITLE
Charlesmchen/websocket state

### PR DESCRIPTION
**Pull request by charlesmchen**
I hit an edge case while testing where the socket manager learned of changes to socket state in `[ensureSocket]`, not from a delegate method. I've reorganized the socket manager to ensure that this case is handled.

PTAL @michaelkirk

### CODE REVIEW

File: src/Messages/Interactions/TSInfoMessage.h [hunk 1]
```diff
@@ -16,8 +17,7 @@ typedef NS_ENUM(NSInteger, TSInfoMessageType) {
     TSInfoMessageTypeDisappearingMessagesUpdate
 };
 
-+ (instancetype)userNotRegisteredMessageInThread:(TSThread *)thread
-                                     transaction:(YapDatabaseReadWriteTransaction *)transaction;
++ (instancetype)userNotRegisteredMessageInThread:(TSThread *)thread;
```
Comment (charlesmchen) [inline src/Messages/Interactions/TSInfoMessage.h:20]: This is some minor unrelated cleanup.  I noticed that we were needlessly passing around a transaction to this method.

File: src/Network/WebSockets/TSSocketManager.m [hunk 2]
```diff
@@ -109,40 +109,132 @@ - (void)ensureWebsocket
...
 
     // Try to reuse the existing socket (if any) if it is in a valid state.
-    SRWebSocket *socket = self.websocket;
-    if (socket) {
-        switch ([socket readyState]) {
+    if (self.websocket) {
+        switch ([self.websocket readyState]) {
             case SR_OPEN:
                 DDLogVerbose(@"WebSocket already open on connection request");
-                OWSAssert(self.status == kSocketStatusOpen);
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m:117]: We can't assume socket and class state are aligned in `ensureWebsocket` - indeed, we need to do the work to ensure they are aligned.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

File: src/Network/WebSockets/TSSocketManager.m [hunk 3]
```diff
@@ -109,40 +109,132 @@ - (void)ensureWebsocket
...
     self.status = kSocketStatusConnecting;
+}
 
-    // Create a new web socket.
-    NSString *webSocketConnect = [textSecureWebSocketAPI stringByAppendingString:[self webSocketAuthenticationString]];
-    NSURL *webSocketConnectURL   = [NSURL URLWithString:webSocketConnect];
-    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:webSocketConnectURL];
+- (NSString *)stringFromSocketStatus:(SocketStatus)status {
+    
+    // If this status update is _not_ redundant,
```
Comment (michaelkirk) [inline src/Network/WebSockets/TSSocketManager.m]: This comment seems misplaced - isn't this just a helper method to pretty-print an enum for logging purposes?

File: src/Messages/Interactions/TSInfoMessage.m [hunk 4]
```diff
@@ -44,8 +45,7 @@ - (instancetype)initWithTimestamp:(uint64_t)timestamp
     return self;
 }
 
-+ (instancetype)userNotRegisteredMessageInThread:(TSThread *)thread
-                                     transaction:(YapDatabaseReadWriteTransaction *)transaction {
++ (instancetype)userNotRegisteredMessageInThread:(TSThread *)thread {
```
Comment (michaelkirk) [inline src/Messages/Interactions/TSInfoMessage.m]: nit clang-format, brace should be on newline

File: src/Network/WebSockets/TSSocketManager.m [hunk 5]
```diff
@@ -109,40 +109,132 @@ - (void)ensureWebsocket
...
+// and [didCloseWithCode:...].  These delegate methods are sometimes
+// invoked _after_ web socket state changes, so we sometimes learn
+// about changes to socket state in [ensureWebsocket].  Put another way,
+// it's not safe to assume we'll learn of changes to websocket state
+// in the websocket delegate methods.
+//
+// Therefore, we use the [setStatus:] setter to ensure alignment between
+// websocket state and class state.
+- (void)setStatus:(SocketStatus)status {
+
```
Comment (michaelkirk) [inline src/Network/WebSockets/TSSocketManager.m]: can we assert this is only called on the WS thread?

File: src/Network/WebSockets/TSSocketManager.m [hunk 6]
```diff
@@ -109,40 +109,132 @@ - (void)ensureWebsocket
...
+                                                   selector:@selector(webSocketHeartBeat)
+                                                   userInfo:nil
+                                                    repeats:YES];
+            
+            // Additionally, we want the ping timer to work in the background too.
+            [[NSRunLoop mainRunLoop] addTimer:self.pingTimer forMode:NSDefaultRunLoopMode];
+            
+            [self.reconnectTimer invalidate];
+            self.reconnectTimer = nil;
+            self.didConnectBg   = YES;
```
Comment (michaelkirk) [inline src/Network/WebSockets/TSSocketManager.m]: nit: Looks like this `didConnectBg` is a vestige of a bygone era. Wanna get rid of it as long as we're tearing this method apart?

**Review by michaelkirk — APPROVED** [review]
LGTM

File: src/Network/WebSockets/TSSocketManager.m (hunk 3 above)
Comment (charlesmchen) [reply src/Network/WebSockets/TSSocketManager.m]: Remove.

File: src/Network/WebSockets/TSSocketManager.m (hunk 5 above)
Comment (charlesmchen) [reply src/Network/WebSockets/TSSocketManager.m]: Actually, it should only be called on the main thread.  I'm going to add asserts to clarify how threading works throughout this file.

File: src/Network/WebSockets/TSSocketManager.m (hunk 6 above)
Comment (charlesmchen) [reply src/Network/WebSockets/TSSocketManager.m]: Done. I went further and removed all vestigial code in this file.

File: src/Messages/Interactions/TSInfoMessage.m (hunk 4 above)
Comment (charlesmchen) [reply src/Messages/Interactions/TSInfoMessage.m]: Done.
